### PR TITLE
Create eps_additon.ml

### PR DESCRIPTION
**Pull request by sjjs7**
Finished proof (final theorem begins line 1732). 
Not ready to commit yet...just for reference.

### CODE REVIEW

File: Constructions/eps_additon.ml [hunk 1]
```diff
@@ -0,0 +1,2136 @@
+needs "Constructions/addition.ml";;
+needs "Constructions/QuotationTactics.ml";;
+
+(*               *)
+(* Term builders *)
+(*               *)
+
+let one_def = define `(One = S Zero)`;;
+
+let thenZero = define `
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:10]: I think it's going to be worth heavily commenting this file, with what each piece means.  For example, although I can read what `thenZero` and `thenOne` do (though a human-math comment would be nice too), I have no idea why you would need these odd functions. That's definitly worth a comment!

File: Constructions/eps_additon.ml [hunk 2]
```diff
@@ -0,0 +1,2136 @@
...
+
+(*                                                                *)
+(* Classify constructions representing proper numbers of type nat *)
+(*                                                                *)
+
+let start_with_one = define `
+  (start_with_one (QuoConst str ty) <=> 
+  	((str = "One") /\ (ty = (TyBase "nat")))) /\
+  (start_with_one (App eps1 eps2) <=> 
+    (((eps1 = (QuoConst "thenZero" (TyBiCons "fun" (TyBase "nat") (TyBase "nat")))) \/
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:28]: I see this `QuotConst...` repeated often. It should be named and pulled out.

File: Constructions/eps_additon.ml [hunk 3]
```diff
@@ -0,0 +1,2136 @@
...
+  (start_with_one (Abs eps1 eps2) = F) /\
+  (start_with_one (QuoVar str ty) = F) /\
+  (start_with_one (Quo e) = F)`;;
+
+let proper_nat_construct = define `
+  (proper_nat_construct (QuoConst str ty) <=> 
+    (((str = "Zero") \/ (str = "One")) /\ (ty = (TyBase "nat")))) /\
+  (proper_nat_construct (App eps1 eps2) <=> 
+  	(((eps1 = (QuoConst "thenZero" (TyBiCons "fun" (TyBase "nat") (TyBase "nat")))) \/
+  	(eps1 = (QuoConst "thenOne" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))))) /\
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:40]: same here.

File: Constructions/eps_additon.ml [hunk 4]
```diff
@@ -0,0 +1,2136 @@
...
+  (proper_nat_construct (Quo eps) = F) /\
+  (proper_nat_construct (Abs eps1 eps2) = F) /\
+  (proper_nat_construct (QuoVar str ty) = F)` ;;
+
+(*                                                                *)
+(* Addition of epsilon numbers                                    *)
+(*                                                                *)
+
+let add_ebin = define `
+  (add_ebin (QuoConst "Zero" (TyBase "nat")) x = x) /\
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:52]: same with all the `QuoConst` that appear here.

File: Constructions/eps_additon.ml [hunk 5]
```diff
@@ -0,0 +1,2136 @@
...
+    (add_ebin x y))) /\
+  (add_ebin (App (QuoConst "thenOne" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))) x) 
+  	(App (QuoConst "thenOne" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))) y) =
+    (App (QuoConst "thenZero" (TyBiCons "fun" (TyBase "nat") (TyBase "nat"))) 
+    (add_ebin (QuoConst "One" (TyBase "nat")) (add_ebin x y))))`;; 
+
+
+(* Lemmas regarding constructions of nats *)
+
+let remove_one = prove(
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:92]: So these next 2 proofs are "the same" except for the rewrites -- create a tactical with the shape, and rewrites are parameters?

File: Constructions/eps_additon.ml [hunk 6]
```diff
@@ -0,0 +1,2136 @@
...
+    ==> start_with_one (add_ebin a1 y)`)) THEN 
+  REWRITE_TAC[ASSUME `start_with_one a1`; ASSUME `start_with_one a1'`] THEN 
+  DISCH_TAC THEN 
+  MP_TAC(SPEC `add_ebin (a1:epsilon) (a1':epsilon)` lemma3) THEN 
+  REWRITE_TAC[ASSUME `start_with_one (add_ebin a1 a1')`]
+  ]
+  ]
+  );;
+
+let lemma6 = prove(
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:370]: Wow, there is serious pain in this `lemma6`!  This is definitely prime territory for abstraction.  You know the 'shape' of `proper_nat_construct`, so there will be patterns-of-proof too.

File: Constructions/eps_additon.ml [hunk 7]
```diff
@@ -0,0 +1,2136 @@
...
+    ASSUME `start_with_one (add_ebin a1 a1')`; 
+    ASSUME `proper_nat_construct (add_ebin a1 a1')`] THEN 
+  REPEAT DISCH_TAC THEN 
+  ASM_REWRITE_TAC[]
+  ]
+  ] 
+  );;
+
+  
+let thm1 = prove(
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:470]: This, one the other hand, seems nice.

File: Constructions/eps_additon.ml [hunk 8]
```diff
@@ -0,0 +1,2136 @@
...
+  (PROPER_NOT_FREE_TAC `(App a0 a1):epsilon` `"j"` `TyBase "epsilon"`) THEN 
+  (PROPER_NOT_FREE_TAC `(add_ebin (App a0 a1) y):epsilon` `"j"` `TyBase "epsilon"`) THEN 
+  NAT_BETA_EVAL_RED(`(\j:epsilon. (eval (j) to (nat))) (y:epsilon)`) THEN 
+  NAT_BETA_EVAL_RED(`(\j:epsilon. (eval (j) to (nat))) ((App a0 a1):epsilon)`) THEN 
+  NAT_BETA_EVAL_RED(`(\j:epsilon. (eval (j) to (nat))) ((add_ebin (App a0 a1) y):epsilon)`) THEN 
+  REPEAT DISCH_TAC THEN 
+  ASM_REWRITE_TAC[]
+  );;
+
+let fin_beta_red9 = prove(
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:847]: probably really worth analyzing the commonalities of these 9 `beta_red` proofs to try to pull out re-usable lemmas!

File: Constructions/eps_additon.ml [hunk 9]
```diff
@@ -0,0 +1,2136 @@
...
+
+(* Needed to do first induction *)
+let ind_1 = NOT_FREE_ABS_NOT_EFFECTIVE_CONV `x:epsilon` 
+  `!y:epsilon. (((proper_nat_construct (x:epsilon)) /\ 
+  (proper_nat_construct (y:epsilon))) ==> 
+  (eval (add_ebin x y) to (nat) = 
+  add_unary (eval x to (nat)) (eval y to (nat))))` `w:epsilon`;;
+addNotEff ind_1;;
+
+let a0_ne = prove(
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:888]: I know why you've called this `a0_ne`, but it's perhaps a sub-optimal name?

File: Constructions/eps_additon.ml [hunk 10]
```diff
@@ -0,0 +1,2136 @@
...
+  let tm = 
+    `(!x:epsilon y:epsilon. 
+    (((proper_nat_construct x) /\ (proper_nat_construct y)) 
+    ==> (eval (add_ebin x y) to (nat) =  
+    add_unary (eval x to (nat)) (eval y to (nat)))))` 
+  in 
+  prove(tm, 
+  MATCH_MP_TAC(lth) THEN 
+  CONJ_TAC THENL
+  [(* isVar x *)
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:1742]: Excellent thing to comment the branches!

File: Constructions/eps_additon.ml [hunk 11]
```diff
@@ -0,0 +1,2136 @@
...
+
+
+
+(*                                                     *)
+(* Meaning formula for addition                        *)
+(*                                                     *)
+
+let add_meaning = 
+  let tm = 
+    `(!x:epsilon y:epsilon.
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:1734]: This term embodies the shape idea I was mentioning yesterday.  Let T be a meta-variable for a type, F a binary syntax transformer and G a binary function (that are supposed to correspond), then your term has shape
```
`(!x:epsilon y:epsilon.
   (((proper_X_construct x) /\ (proper_X_construct y))
   ==> (eval (F x y) to (X) =
   G (eval x to (X)) (eval y to (X))))))`
```
This also points out that putting X in the name of `proper_construct` may be limiting.  @wmfarmer should ponder on this too.

File: Constructions/eps_additon.ml [hunk 12]
```diff
@@ -0,0 +1,2136 @@
...
+  ]
+  ;REWRITE_TAC[proper_nat_construct]
+  ]
+  ]
+  ]
+  ;REWRITE_TAC[proper_nat_construct]
+  ]
+  ]
+  ]
+  );;
```
Comment (JacquesCarette) [inline Constructions/eps_additon.ml:2133]: That is indeed quite the monster. I definitely think it should be split into lemmas, corresponding to the shape that I indicated about in the comment above, to isolate the shape-driven tactics from the nat-specific parts, as much as possible.

**Review by JacquesCarette — COMMENTED** [review]
Very impressive!

File: Constructions/eps_additon.ml (hunk 2 above)
Comment (sjjs7) [reply Constructions/eps_additon.ml:28]: What would you say is the best way to do this? I can't seem to find an `as` keyword for definitions in HOL light like there is in OCaml and when I tried to make definitions (ex. let ZERO = new_definition `ZERO = QuoConst "Zero" (TyBase "nat")`  which creates a theorem) that caused more problems.

Comment (JacquesCarette) [reply Constructions/eps_additon.ml:28]: I think you want to make those definitions be at the OCaml level, not the HOL Light level.

**Pull request closed without merging**